libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp

```diff
@@ -0,0 +1,286 @@
+    crc += BLPING_DEST_ID;
+
+    // payload
+    if (payload != nullptr) {
+        for (uint16_t i = 0; i<payload_len; i++) {
+            uart->write(payload[i]);
+            crc += payload[i];
+        }
+    }
+
+    // checksum
+    uart->write(LOWBYTE(crc));
+    uart->write(HIGHBYTE(crc));
+}
+
+// distance returned in reading_cm, signal_ok is set to true if sensor reports a strong signal
+bool AP_RangeFinder_BLPing::get_reading(uint16_t &reading_cm)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    float sum_cm = 0;
+    uint16_t count = 0;
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
+        if (b < 0) {
+            continue;
+        }
+        if (parse_byte(b)) {
+            count++;
+            sum_cm += distance_cm;
+        }
+    }
+
+    if (count > 0) {
+        // return average distance of readings
+        reading_cm = sum_cm / count;
+
+        // request another distance
+        send_message(BLDPIN_MSGID_DISTANCE_SIMPLE, nullptr, 0);
+
+        return true;
+    }
+
+    // no readings so return false
+    return false;
+}
+
+// process one byte received on serial port
+// returns true if a distance message has been successfully parsed
+// state is stored in msg structure
+bool AP_RangeFinder_BLPing::parse_byte(uint8_t b)
+{
+    bool got_distance = false;
+
+    // process byte depending upon current state
+    switch (msg.state) {
+
+    case MSG_STATE_HEADER1:
+        if (b == BLPING_FRAME_HEADER1) {
+            msg.crc_expected = BLPING_FRAME_HEADER1;
+            msg.state = MSG_STATE_HEADER2;
+        }
+        break;
+
+    case MSG_STATE_HEADER2:
+        if (b == BLPING_FRAME_HEADER2) {
+            msg.crc_expected += BLPING_FRAME_HEADER2;
+            msg.state = MSG_STATE_LEN_L;
+        } else {
+            msg.state = MSG_STATE_HEADER1;
+        }
+        break;
+
+    case MSG_STATE_LEN_L:
+        msg.payload_len = b;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_LEN_H;
+        break;
+
+    case MSG_STATE_LEN_H:
+        msg.payload_len += ((uint16_t)b << 8);
+        msg.payload_recv = 0;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_MSG_ID_L;
+        break;
+
+    case MSG_STATE_MSG_ID_L:
+        msg.msgid = b;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_MSG_ID_H;
+        break;
+
+    case MSG_STATE_MSG_ID_H:
+        msg.msgid += ((uint16_t)b << 8);
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_SRC_ID;
+        break;
+
+    case MSG_STATE_SRC_ID:
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_DST_ID;
+        break;
+
+    case MSG_STATE_DST_ID:
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_PAYLOAD;
+        break;
+
+    case MSG_STATE_PAYLOAD:
+        if (msg.payload_recv < msg.payload_len) {
+            msg.payload[msg.payload_recv] = b;
```

Comment:
This looks like  a nasty buffer overwrite issue; `msg.payload[]` is only 20 bytes long ATM.  Suggest just throwing away bytes if `msg.payload_recv >= ARRAY_SIZE(msg.payload)`

Reply:
added check before adding new character to the msg structure's payload array.